Detectors/Vertexing/src/SVertexer.cxx

```diff
@@ -1090,60 +1096,68 @@ int SVertexer::check3bodyDecays(const V0Index& v0Idx, const V0& v0, float rv0, s
     tr0.getPxPyPzGlo(p0);
     tr1.getPxPyPzGlo(p1);
     tr2.getPxPyPzGlo(p2);
-    std::array<float, 3> p3B = {p0[0] + p1[0] + p2[0], p0[1] + p1[1] + p2[1], p0[2] + p1[2] + p2[2]};
 
-    float pt2candidate = p3B[0] * p3B[0] + p3B[1] * p3B[1], p2candidate = pt2candidate + p3B[2] * p3B[2];
-    if (pt2candidate < mMinPt23Body) { // pt cut
-      continue;
-    }
-    if (p3B[2] * p3B[2] / pt2candidate > mMaxTgl23Body) { // tgLambda cut
-      continue;
-    }
-
-    // compute primary vertex and cosPA of the 3-body decay
-    auto bestCosPA = mSVParams->minCosPA3body;
+    bool goodHyp = false;
+    o2::track::PID pidHyp = o2::track::PID::Electron; // Update if goodHyp is true
     auto decay3bodyVtxID = -1;
+    auto vtxCosPA = -1;
+
+    std::array<float, 3> pbach = {0, 0, 0}, p3B = {0, 0, 0}; // Update during the check of invariant mass
+    for (int ipid = 0; ipid < NHyp3body; ipid++) {
+      // check mass based on hypothesis of charge of bachelor (pos and neg expected to be proton/pion)
+      float bachChargeFactor = m3bodyHyps[ipid].getChargeBachProng() / tr2.getAbsCharge();
+      pbach = {bachChargeFactor * p2[0], bachChargeFactor * p2[1], bachChargeFactor * p2[2]};
+      p3B = {p0[0] + p1[0] + pbach[0], p0[1] + p1[1] + pbach[1], p0[2] + p1[2] + pbach[2]};
+      float sqP0 = p0[0] * p0[0] + p0[1] * p0[1] + p0[2] * p0[2], sqP1 = p1[0] * p1[0] + p1[1] * p1[1] + p1[2] * p1[2], sqPBach = pbach[0] * pbach[0] + pbach[1] * pbach[1] + pbach[2] * pbach[2];
+      float pt2candidate = p3B[0] * p3B[0] + p3B[1] * p3B[1], p2candidate = pt2candidate + p3B[2] * p3B[2];
+      float ptCandidate = std::sqrt(p3B[0] * p3B[0] + p3B[1] * p3B[1]);
+      if (m3bodyHyps[ipid].check(sqP0, sqP1, sqPBach, p2candidate, ptCandidate)) {
+        if (pt2candidate < mMinPt23Body) { // pt cut
+          continue;
+        }
+        if (p3B[2] * p3B[2] / pt2candidate > mMaxTgl23Body) { // tgLambda cut
```

Comment:
it is faster to do `if (p3B[2] * p3B[2] > pt2candidate * mMaxTgl23Body) {`

Reply:
Fixed